mim/commands/install.py

```diff
@@ -1,580 +1,306 @@
+    """A patch for `pip._vendor.pkg_resources.Distribution`.
 
+    Since the old version of the OpenMMLab packages did not add the 'mim' extra
+    requirements to the release distribution, we need to hack the Distribution
+    and manually fetch the 'mim' requirements from `mminstall.txt`.
 
-def parse_dependencies(path: str) -> list:
-    """Parse dependencies from repo/requirements/mminstall.txt.
+    This patch works with 'pip<22.1' and 'pip>=22.1,python<3.11'.
 
     Args:
-        path (str): Path of mminstall.txt.
+        index_url (str, optional): The pypi index url that pass to
+            `get_mmdeps_from_mmpkg_pypi`.
     """
+    from pip._vendor.pkg_resources import Distribution, parse_requirements
 
-    def _get_proper_version(package, version, op):
-        releases = get_release_version(package)
-        if op == '>':
-            for r_v in releases:
-                if LooseVersion(r_v) > LooseVersion(version):
-                    return r_v
-            else:
-                raise ValueError(
-                    highlighted_error(f'invalid min version of {package}'))
-        elif op == '<':
-            for r_v in releases[::-1]:
-                if LooseVersion(r_v) < LooseVersion(version):
-                    return r_v
-            else:
-                raise ValueError(
-                    highlighted_error(f'invalid max version of {package}'))
-
-    dependencies = []
-    with open(path) as fr:
-        for requirement in parse_requirements(fr):
-            pkg_name = requirement.project_name
-            min_version = ''
-            max_version = ''
-            for op, version in requirement.specs:
-                if op == '==':
-                    min_version = max_version = version
-                    break
-                elif op == '>=':
-                    min_version = version
-                elif op == '>':
-                    min_version = _get_proper_version(pkg_name, version, '>')
-                elif op == '<=':
-                    max_version = version
-                elif op == '<':
-                    max_version = _get_proper_version(pkg_name, version, '<')
-
-            dependencies.append([pkg_name, min_version, max_version])
-
-    return dependencies
-
-
-def install_dependencies(dependencies: List[List[str]],
-                         timeout: int = 15,
-                         is_yes: bool = False,
-                         is_user_dir: bool = False) -> None:
-    """Install dependencies, such as mmcls depends on mmcv.
+    origin_requires = Distribution.requires
+
+    def patched_requires(self, extras=()):
+        deps = origin_requires(self, extras)
+        if self.project_name not in PKG2PROJECT or self.project_name == 'mmcv-full':  # noqa: E501
+            return deps
+
+        if 'mim' in self.extras:
+            mim_extra_requires = origin_requires(self, ('mim', ))
+            filter_invalid_marker(mim_extra_requires)
+            deps += mim_extra_requires
```

Comment:
What is the type of `deps`? If it is list, we can use `deps.extend(mim_extra_requires)`.

Reply:
Why use the extend method instead of +=, is there any difference or is the programming style required to do so.

Reply:
I prefer to extend but `+=` maybe has a better performance so `+=` is okay.